sql: panic when adding a default value that is assignment cast to the required type

**Describe the problem**

The latest version of the CockroachDB (dc93fa8ccfc0) crashes when executing the following query:

```sql
CREATE TABLE t (c1 INT);
INSERT INTO t VALUES (0);
ALTER TABLE t ADD COLUMN c2 DECIMAL DEFAULT pi();
SELECT * FROM t;
```

**To Reproduce**

Here is the detail steps to reproduce the bug.

1. In system Ubuntu 20.04, download the CockroachDB source code from the github source.
2. Use the latest version of the CockroachDB code (tested version: dc93fa8ccfc0)
3. Directly `make install` in the root repository folder. 
4. Run `./cockroach demo`, and then paste the PoC query to the cockroach cli environment.
5. Observe the crashing and log the crashing information. 

**Expected behavior**
The SELECT statement should be executed correctly and it should return matched rows. 

**Additional data / screenshots**

Here is the crashing stack trace: 

```
*
* ERROR: a SQL panic has occurred while executing the following statement:
* SELECT ta3.c2 FROM v0 AS ta3 WHERE ta3.c2 < 0
*
*
* ERROR: a panic has occurred!
* runtime error: slice bounds out of range [:64] with capacity 7
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
*   | [...repeated from below...]
* Wraps: (2) while executing: SELECT _._ FROM _ AS _ WHERE _._ < _
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:884
*   | runtime.goPanicSliceAcap
*   | 	/usr/local/go/src/runtime/panic.go:139
*   | github.com/cockroachdb/cockroach/pkg/util/encoding.DecodeIntoUntaggedDecimalValue
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/encoding/encoding.go:2894
*   | github.com/cockroachdb/cockroach/pkg/sql/colencoding.DecodeTableValueToCol
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colencoding/value_encoding.go:72
*   | github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).processValueBytes
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:1180
*   | github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).processValue
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:981
*   | github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).NextBatch
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:760
*   | github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:108
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*CancelChecker).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/cancel_checker.go:59
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:116
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:124
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecsel.(*selLTInt64Int64ConstOp).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecsel/selection_ops.eg.go:24779
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.(*simpleProjectOp).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase/simple_project.go:124
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:116
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:124
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:252
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:256
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:294
*   | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:309
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1768
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll
*   | 	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1520
* Wraps: (4) runtime error: slice bounds out of range [:64] with capacity 7
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.boundsError
*
panic: runtime error: slice bounds out of range [:64] with capacity 7 [recovered]
	panic: runtime error: slice bounds out of range [:64] with capacity 7 [recovered]
	panic: runtime error: slice bounds out of range [:64] with capacity 7 [recovered]
	panic: runtime error: slice bounds out of range [:64] with capacity 7 [recovered]
	panic: runtime error: slice bounds out of range [:64] with capacity 7

goroutine 4894 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc004903000, {0x82aa4d0, 0xc005078c40}, {0x516e3a0?, 0xc00546c7f8?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1080 +0x1b4
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1({0x82aa4d0, 0xc005078c40}, {0xbac2f0e?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:832 +0x4a
panic({0x516e3a0, 0xc00546c7f8})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x407
panic({0x516e3a0, 0xc00546c7f8})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x407
panic({0x516e3a0, 0xc00546c7f8})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x407
panic({0x516e3a0, 0xc00546c7f8})
	/usr/local/go/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/cockroach/pkg/util/encoding.DecodeIntoUntaggedDecimalValue(0xc00546c7ea?, {0xc00546c7ec?, 0xa?, 0xc005896da8?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/util/encoding/encoding.go:2894 +0xe5
github.com/cockroachdb/cockroach/pkg/sql/colencoding.DecodeTableValueToCol(0x0?, 0xc001609978, 0x7fae8a61f9f8?, 0x0, 0xc00453717b?, 0x1?, 0xc0045fbec0?, {0xc00546c7eb, 0x9, 0x9})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colencoding/value_encoding.go:72 +0x2c5
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).processValueBytes(0xc001609880, {0x14?, 0x14?}, 0xc0034ccc00, {0xc00546c7e5, 0xf, 0xf}, {0x0, 0x0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:1180 +0x3ee
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).processValue(0xc001609880, {0x82aa578, 0xc005838f00}, 0x0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:981 +0xea7
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*cFetcher).NextBatch(0xc001609880, {0x82aa578, 0xc005838f00})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/cfetcher.go:760 +0x1029
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).Next(0xc0017197c0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:108 +0x30
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils.(*CancelChecker).Next(0xc001d56b00)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecutils/cancel_checker.go:59 +0x30
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next(0xc0035a3770)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:116 +0x29
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc001d56b40?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next(0xc0035a3770)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:124 +0x52
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecsel.(*selLTInt64Int64ConstOp).Next(0xc005838db0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecsel/selection_ops.eg.go:24779 +0x32
github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.(*simpleProjectOp).Next(0xc004b90720)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase/simple_project.go:124 +0x3f
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).next(0xc0035a3950)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:116 +0x29
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc001d56c00?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next(0xc0035a3950)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:124 +0x52
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).nextAdapter(0xc003224460)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:252 +0x2a
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc00430c000?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x62
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).next(...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:256
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*BatchFlowCoordinator).Run(0xc003224460, {0x82aa4d0?, 0xc001d56ac0?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:294 +0x2e5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0xc00182d2c0, {0x82aa4d0, 0xc001d56ac0})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:309 +0x1d8
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc003828dc0, {0x82aa578, 0xc0058387e0}, 0xc00167ea80, 0xc001dfab00, 0xc001d3e400, 0xc001609c00, 0xc0049038b0, 0x0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:815 +0x984
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc005838870?, {0x82aa578, 0xc0058387e0}, 0x0?, 0xc00167ea80, 0x0?, {{0x82acb50?, 0xc004edb5e0?}, 0x0?}, 0xc001609c00)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1768 +0x1f9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0xc00167ea80?, {0x82aa578, 0xc0058387e0}, 0xc004903540?, 0x1?, 0xc004903540, 0xc001609c00, 0xc003b31d40?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1520 +0x206
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc004903000, {0x82aa578, 0xc0058387e0}, 0xc004903540, 0xc0058387e0?, {0x7fae8a79ac00?, 0xc003dacb10}, 0x0?, 0xc004099ad8)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1624 +0x4c8
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc004903000, {0x82aa578, 0xc005838000}, 0xc004903540, {0x7fae8a79ac00, 0xc003dacb10})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1245 +0xc13
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc004903000, {0x82aa578?, 0xc0051fbf80?}, {{0x82d3800, 0xc00444bb30}, {0x0, 0x0, 0x0}, {0xc0042085a8, 0x2d}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:737 +0x2634
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x82aa578?, 0xc0051fbf80?})
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:130 +0xbc
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x82aa578?, {0x82aa578?, 0xc0051fbf80?}, {0x82d3800?, 0xc00444bb30?}, 0x0?, 0xaf47700?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2513 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc004903000, {0x82aa578, 0xc0051fbf80}, {{0x82d3800, 0xc00444bb30}, {0x0, 0x0, 0x0}, {0xc0042085a8, 0x2d}, ...}, ...)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:129 +0x4fa
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0xc005899ac0, 0xc004903000, 0xc0058997b8, 0xc005899870, 0xc0058998b0, 0xc005899890, 0xc005899880, 0xc0058998a0)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1955 +0x30f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc004903000)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1960 +0x1127
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc004903000, {0x82aa4d0, 0xc005078c40}, 0x4194ca8ec8000000?, 0xc004cc1dd8?, 0xc004e3dc30?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1878 +0x20b
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc003b2ab40?, {0x82aa4d0?, 0xc005078c40?}, {0xc001c39100?}, 0x3?, 0xc0060f75a0?)
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:834 +0xe6
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1()
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:728 +0x40d
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/home/cockroach/go_projects/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:639 +0x22e
```

**Environment:**
 - CockroachDB version: v23.1.0-alpha
 - Server OS: Ubuntu 20.04 LTS
 - Client app: CockroachDB demo command. (`./cockroach demo`)
 - Detailed CockroachDB version output:

```
Build Tag:        v23.1.0-alpha.00000000-1118-gdc93fa8ccf-dirty
Build Time:       2022/12/10 21:03:15
Distribution:     CCL
Platform:         linux amd64 (x86_64-linux-gnu)
Go Version:       go1.19
C Compiler:       gcc 9.4.0
Build Commit ID:  dc93fa8ccfc0d08afd81a3222426d5608fd7a529
Build Type:       development
```

**Additional context**

The crashing problem may not happen in the real-world application, since it requires the specific setup: `SET testing_optimizer_disable_rule_probability = 1.0;`, which is a developer option used to control the query optimization. However, a crash is still unexpected from the resulted behavior. Therefore, this problem is reported as a crashing bug. 


Jira issue: CRDB-22312

--- a/pkg/sql/catalog/schemaexpr/default_exprs.go
+++ b/pkg/sql/catalog/schemaexpr/default_exprs.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/transform"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/errors"
 )
 
 // MakeDefaultExprs returns a slice of the default expressions for the slice
@@ -70,6 +71,20 @@ func MakeDefaultExprs(
 		typedExpr, err := tree.TypeCheck(ctx, expr, semaCtx, col.GetType())
 		if err != nil {
 			return nil, err
+		}
+		// For "reasons" the CastExpr type check ignores the desired type. That
+		// means that we can get a result here that is not the right type, and
+		// we'd need to wrap that in an explicit cast. This is the equivalent of
+		// an assignment cast we'd insert when writing to the table directly.
+		if !typedExpr.ResolvedType().Equivalent(col.GetType()) {
+			if typedExpr, err = tree.TypeCheck(ctx, &tree.CastExpr{
+				Expr:       typedExpr,
+				Type:       col.GetType(),
+				SyntaxMode: tree.CastExplicit,
+			}, semaCtx, col.GetType()); err != nil {
+				return nil, errors.NewAssertionErrorWithWrappedErrf(err,
+					"failed to type check the cast of %v to %v", expr, col.GetType())
+			}
 		}
 		if typedExpr, err = txCtx.NormalizeExpr(ctx, evalCtx, typedExpr); err != nil {
 			return nil, err

--- a/pkg/sql/logictest/testdata/logic_test/alter_table
+++ b/pkg/sql/logictest/testdata/logic_test/alter_table
@@ -2751,3 +2751,22 @@ SELECT count(*) from t_with_dropped_index_expr;
 
 statement ok
 RESUME JOB (SELECT job_id FROM crdb_internal.jobs WHERE description LIKE 'ALTER TABLE %t_with_dropped_index_expr DROP COLUMN j CASCADE%' AND status='paused' FETCH FIRST 1 ROWS ONLY);
+
+
+# In 23.1 we added support for default expressions which can be assignment cast
+# to the column type. This work introduced a bug whereby the backfill logic
+# would not apply the appropriate cast. This test ensures that such tables can
+# be backfilled.
+subtest add_column_with_default_expression_with_different_type
+
+statement ok
+CREATE TABLE t_93398 (c1 INT);
+INSERT INTO t_93398 VALUES (0);
+
+statement ok
+ALTER TABLE t_93398 ADD COLUMN c2 DECIMAL DEFAULT pi();
+
+query IT
+SELECT * FROM t_93398;
+----
+0  3.141592653589793

